csr_check_test.go

```diff
@@ -1940,49 +2111,241 @@ func Test_authorizeCSR(t *testing.T) {
+								},
+								{
+									Type:    corev1.NodeInternalDNS,
+									Address: "node1.local",
+								},
+								{
+									Type:    corev1.NodeExternalDNS,
+									Address: "node1",
+								},
+							},
+						},
+					},
+				},
+				req: &certificatesv1beta1.CertificateSigningRequest{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              "renew",
+						CreationTimestamp: creationTimestamp(10 * time.Minute),
+					},
+					Spec: certificatesv1beta1.CertificateSigningRequestSpec{
+						Usages: []certificatesv1beta1.KeyUsage{
+							certificatesv1beta1.UsageKeyEncipherment,
+							certificatesv1beta1.UsageDigitalSignature,
+							certificatesv1beta1.UsageServerAuth,
+						},
+						Username: "system:node:test",
+						Groups: []string{
+							"system:authenticated",
+							"system:nodes",
+						},
+					},
+				},
+				csr: goodCSR,
+				ca:  []*x509.Certificate{parseCert(t, rootCertGood)},
+			},
+		},
+		{
+			name: "successfull fallback to fresh approval from incorrect server cert",
+			args: args{
+				node: withPort(defaultPort+1, withName("test", defaultNode())),
+				req: &certificatesv1beta1.CertificateSigningRequest{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              "renew",
+						CreationTimestamp: creationTimestamp(10 * time.Minute),
+					},
+					Spec: certificatesv1beta1.CertificateSigningRequestSpec{
+						Usages: []certificatesv1beta1.KeyUsage{
+							certificatesv1beta1.UsageKeyEncipherment,
+							certificatesv1beta1.UsageDigitalSignature,
+							certificatesv1beta1.UsageServerAuth,
+						},
+						Username: "system:node:test",
+						Groups: []string{
+							"system:authenticated",
+							"system:nodes",
+						},
+					},
+				},
+				csr:           goodCSR,
+				ca:            []*x509.Certificate{parseCert(t, differentCert)},
+				kubeletServer: fakeResponder(t, fmt.Sprintf("%s:%v", defaultAddr, defaultPort+1), differentCert, differentKey),
+			},
+			wantErr: `No target machine for node "test"`,
 		},
 	}
+
+	server := fakeResponder(t, fmt.Sprintf("%s:%v", defaultAddr, defaultPort), serverCertGood, serverKeyGood)
+	defer server.Close()
+
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			kubeletServer := server
+			if tt.args.kubeletServer != nil {
+				kubeletServer = tt.args.kubeletServer
+				defer kubeletServer.Close()
+			}
+
+			nodes := &testNode{t: t, node: tt.args.node, err: tt.args.nodeErr}
 			tt.args.req.Spec.Request = []byte(tt.args.csr)
 			parsedCSR, err := parseCSR(tt.args.req)
 			if err != nil {
 				t.Fatal(err)
 			}
-			nodes := &testNode{t: t, name: tt.args.nodeName, err: tt.args.nodeErr}
 
-			if err := authorizeCSR(tt.args.config, tt.args.machines, nodes, tt.args.req, parsedCSR, tt.args.ca); errString(err) != tt.wantErr {
+			var ca *x509.CertPool
+			if len(tt.args.ca) > 0 {
+				// Start renewal flow
+				ca = x509.NewCertPool()
+				for _, cert := range tt.args.ca {
+					ca.AddCert(cert)
+				}
+				go respond(kubeletServer)
```

Comment:
is there any chance we would deadlock here inside of `respond`? (i saw an `accept` call in that function)

Reply:
https://github.com/openshift/cluster-machine-approver/pull/95#discussion_r527597436